File: leetCode/Python/189-rotateArray.py
class Solution1(object):
    def rotate(self, nums, k):
        '''
        TC: O(n)
        SC: O(n)
        '''
        k %= len(nums)
        rotate_len = len(nums) - k
        nums.extend(nums[:rotate_len])
        del nums[:rotate_len]

class Solution(object):
    def rotate(self, nums, k):
        """
        TC: O(n)
        SC: O(1)
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        k %= len(nums)
        self.reverse(nums, 0, len(nums) - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, len(nums) - 1)

    def reverse(self, nums, start, end):
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            end   -= 1
            start += 1

File: leetCode/C++/189-rotateArray.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        if (nums.size() == 0)
            return;
        k %= nums.size();
        reverse(nums, 0, nums.size() - k - 1);
        reverse(nums, nums.size() - k, nums.size() - 1);
        reverse(nums, 0, nums.size() - 1);
    }
private:
    void reverse(vector<int>& nums, int start, int end) {
        while (start < end) {
            int tmp = nums[start];
            nums[start] = nums[end];
            nums[end] = tmp;
            start++;
            end--;
        }
    }
};

int main(void) {
    vector<int> nums = {1, 2, 3, 4, 5, 6, 7};
    Solution s;
    s.rotate(nums, 10);
    for (const auto &num: nums)
        cout << num << " ";
    cout << endl;
};